### PR TITLE
DataFrame.reindex supports koalas Index parameter

### LINKED ISSUE #1740: DataFrame.reindex with a koalas Index parameter is not implemented
In pandas, we have :

```python
>>> df = pd.DataFrame({"a": [1, 2, 3]}, index=["A", "B", "C"])
>>> df2 = pd.DataFrame({"a": [4, 5, 6]}, index=["A", "D", "B"])
>>> df.reindex(index=df2.index)
     a
A  1.0
D  NaN
B  2.0
```

But it is not implemented in koalas :

```python
>>> df = ks.from_pandas(df)
>>> df2 = ks.from_pandas(df2)
>>> df = df.reindex(df2.index)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\Dev\koalas\databricks\koalas\frame.py", line 7805, in reindex
    df = df._reindex_index(index, fill_value)
  File "D:\Dev\koalas\databricks\koalas\frame.py", line 7824, in _reindex_index
    kser = ks.Series(list(index))
  File "D:\Dev\koalas\databricks\koalas\indexes.py", line 1989, in __iter__
    return MissingPandasLikeIndex.__iter__(self)
  File "D:\Dev\koalas\databricks\koalas\missing\__init__.py", line 23, in unsupported_function
    class_name=class_name, method_name=method_name, reason=reason
databricks.koalas.exceptions.PandasNotImplementedError: The method `pd.Index.__iter__()` is not implemented. If you want to collect your data as an NumPy array, use 'to_numpy()' instead.
```

It could be interesting to have this feature, I will submit a PR for this.

### PR DESCRIPTION
**Pull request by LucasG0**
This PR would close #1740.

### CODE REVIEW

File: databricks/koalas/tests/test_dataframe.py [hunk 1]
```diff
@@ -2396,6 +2396,15 @@ def test_reindex(self):
             kdf.reindex(columns=["numbers", "2", "3"], fill_value=0.0).sort_index(),
         )
 
+        index2 = ["A", "C", "D", "E", "0"]
+        pdf2 = pd.DataFrame({"a": [1.0, 2.0, 3.0, 4.0, 5.0]}, index=index)
```
Comment (itholic) [inline databricks/koalas/tests/test_dataframe.py:2400]: maybe `index=index2` ?

Comment (itholic) [inline databricks/koalas/tests/test_dataframe.py:2400]: Maybe `index=index2` ?

**Review by itholic — APPROVED** [review]
Otherwise, seems fine to me.

File: databricks/koalas/tests/test_dataframe.py (hunk 1 above)
Comment (LucasG0) [reply databricks/koalas/tests/test_dataframe.py]: Sure thanks !

File: databricks/koalas/tests/test_dataframe.py [hunk 2]
```diff
@@ -2396,6 +2396,15 @@ def test_reindex(self):
             kdf.reindex(columns=["numbers", "2", "3"], fill_value=0.0).sort_index(),
         )
 
+        index2 = ["A", "C", "D", "E", "0"]
```
Comment (itholic) [inline databricks/koalas/tests/test_dataframe.py:2399]: Oh, could add test with named index?

like

```python
index2 = pd.Index(["A", "C", "D", "E", "0"], name="index")
```

I think maybe It wouldn't work ?

Comment (HyukjinKwon) [reply databricks/koalas/tests/test_dataframe.py:2399]: Oh yeah, good point.

File: databricks/koalas/tests/test_dataframe.py [hunk 3]
```diff
@@ -2396,6 +2396,15 @@ def test_reindex(self):
...
+        pdf2 = pd.DataFrame({"a": [1.0, 2.0, 3.0, 4.0, 5.0]}, index=index2)
+        kdf2 = ks.from_pandas(pdf2)
+
+        self.assert_eq(
+            pdf.reindex(index=pdf2.index, fill_value=0.0).sort_index(),
+            kdf.reindex(index=kdf2.index, fill_value=0.0).sort_index(),
+        )
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:2406]: Also what if the given index is `MultiIndex`?

File: databricks/koalas/tests/test_dataframe.py [hunk 4]
```diff
@@ -2396,6 +2396,15 @@ def test_reindex(self):
             kdf.reindex(columns=["numbers", "2", "3"], fill_value=0.0).sort_index(),
         )
 
+        index2 = ["A", "C", "D", "E", "0"]
+        pdf2 = pd.DataFrame({"a": [1.0, 2.0, 3.0, 4.0, 5.0]}, index=index2)
+        kdf2 = ks.from_pandas(pdf2)
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:2401]: I guess we can just use `Index`?

```py
pindex2 = pd.Index(["A", "C", "D", "E", "0"])
kindex2 = ks.from_pandas(pindex2)
```

**Comment by LucasG0** [comment]
I updated with named `Index` + `MultiIndex` support.
For` MultiIndex`, as currently only reindexing one level index is supported, 
it just returns a filled `DataFrame` with new index and current object columns.
Also, we can now reindex columns of an empty `DataFrame` as pandas supports.

File: databricks/koalas/tests/test_dataframe.py (hunk 3 above)
Comment (LucasG0) [reply databricks/koalas/tests/test_dataframe.py:2432]: Thanks for noticing, I updated PR so it has the same behavior of pandas, considering we currently can only reindex single level index.

File: databricks/koalas/frame.py [hunk 5]
```diff
@@ -7851,7 +7864,7 @@ def _reindex_index(self, index, fill_value):
         else:
             joined_df = frame.join(labels, on=index_column, how="right")
 
-        internal = self._internal.with_new_sdf(joined_df)
+        internal = InternalFrame(joined_df, index_map=index_map)
```
Comment (ueshin) [inline databricks/koalas/frame.py:7867]: This removes `column_labels` and doesn't drop `NATURAL_ORDER_COLUMN_NAME`. Also we should update `data_spark_columns`.
Could you follow what's done in `InternalFrame. with_new_sdf`.

File: databricks/koalas/frame.py [hunk 6]
```diff
@@ -7863,11 +7876,12 @@ def _reindex_columns(self, columns, fill_value):
                     raise TypeError("Expected tuple, got {}".format(type(col)))
         else:
             label_columns = [(col,) for col in columns]
-        for col in label_columns:
-            if len(col) != level:
-                raise ValueError(
-                    "shape (1,{}) doesn't match the shape (1,{})".format(len(col), level)
-                )
+        if level != 0:
```
Comment (ueshin) [inline databricks/koalas/frame.py:7879]: Do we need this?

File: databricks/koalas/frame.py (hunk 5 above)
Comment (LucasG0) [reply databricks/koalas/frame.py:7867]: Yes, I am working on it

**Comment by LucasG0** [comment]
Hi, I updated PR to fully support MultiIndex.
To be consistent with pandas behavior :
- Reindexing MultiIndex on multiIndexed DataFrame is supported
- Reindexing single Index on multiIndexex DataFrame raise an error
- Reindexing MultiIndex on single indexed DataFrame is supported and returns a DataFrame with new index and original columns filled with fill_value

File: databricks/koalas/frame.py [hunk 7]
```diff
@@ -7830,43 +7830,89 @@ def reindex(
...
+            )
+
         if fill_value is not None:
-            frame_index_column = verify_temp_column_name(frame, "__frame_index_column__")
-            frame = frame.withColumnRenamed(index_column, frame_index_column)
+            frame_index_columns = [
+                verify_temp_column_name(frame, "__frame_index_column_{}__".format(i))
+                for i in range(nlevels)
+            ]
+            cols = [F.col(t[0]).alias(t[1]) for t in zip(index_columns, frame_index_columns)]
```
Comment (ueshin) [inline databricks/koalas/frame.py:7869]: `scol_for(frame, t[0])` instead of `F.col(t[0])`?

Comment (ueshin) [reply databricks/koalas/frame.py:7869]: Btw, we prefer to use a tuple, e.g., `[... for index_col, frame_index_col in zip(index_columns, frame_index_columns)]`.

**Review by ueshin — COMMENTED** [review]
Sorry for the delay. I left some comments. Thanks!

File: databricks/koalas/frame.py [hunk 8]
```diff
@@ -7830,43 +7830,89 @@ def reindex(
...
+                return DataFrame(
+                    index._internal.with_new_columns(
+                        [F.lit(0)],
+                        column_labels=[(verify_temp_column_name(frame, "__temp_spark_column__"),)],
+                    )
+                ).reindex(columns=self._internal.data_spark_column_names, fill_value=fill_value)
+
+            obj = index
+            index_map = OrderedDict(
+                [(SPARK_INDEX_NAME_FORMAT(i), (name,)) for i, name in enumerate(index.names)]
```
[the remark this answers is not in the document]
Comment (LucasG0) [reply databricks/koalas/frame.py]: I could have but now I do not use `index._internal.index_map` to handle when index and frame have different `index_spark_column_names`.

File: databricks/koalas/frame.py [hunk 9]
```diff
@@ -7830,43 +7830,89 @@ def reindex(
...
+            )
+            labels = obj._internal.spark_frame.select(index_columns)
```
[the remark this answers is not in the document]
Comment (LucasG0) [reply databricks/koalas/frame.py]: In deed thanks for pointing out !

File: databricks/koalas/frame.py [hunk 10]
```diff
@@ -7830,43 +7830,89 @@ def reindex(
...
         else:
-            joined_df = frame.join(labels, on=index_column, how="right")
+            joined_df = frame.join(labels, on=index_columns, how="right")
 
-        internal = self._internal.with_new_sdf(joined_df)
+        sdf = joined_df.drop(NATURAL_ORDER_COLUMN_NAME)
+        internal = InternalFrame(
+            sdf,
+            index_map=index_map,
+            column_labels=self._internal._column_labels,
```
Comment (ueshin) [inline databricks/koalas/frame.py:7910]: nit: `self._internal.column_labels`? We'd prefer to use the property without starting with `_`.

File: databricks/koalas/frame.py [hunk 11]
```diff
@@ -7830,43 +7830,89 @@ def reindex(
...
+            data_spark_columns=[
+                scol_for(sdf, col) for col in self._internal.data_spark_column_names
+            ],
+            column_label_names=self._internal._column_label_names,
```
Comment (ueshin) [inline databricks/koalas/frame.py:7914]: ditto.

File: databricks/koalas/frame.py [hunk 12]
```diff
@@ -7830,43 +7830,86 @@ def reindex(
...
-        index_column = self._internal.index_spark_column_names[0]
-
-        kser = ks.Series(list(index))
-        labels = kser._internal.spark_frame.select(kser.spark.column.alias(index_column))
+        index_columns = self._internal.index_spark_column_names
         frame = self._internal.resolved_copy.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME)
 
+        if isinstance(index, ks.Index):
+            if nlevels != index.nlevels:
+                # Use temporary column so we can call reindex in order to fill columns on new index
```
Comment (ueshin) [inline databricks/koalas/frame.py:7843]: nit: do you want to remove or rephrase this comment?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: databricks/koalas/frame.py (hunk 12 above)
Comment (LucasG0) [reply databricks/koalas/frame.py:7843]: Yes I removed it

**Comment by ueshin** [comment]
Thanks! merging.

**Comment by LucasG0** [comment]
Thanks ! :)

**Pull request merged**